### PR TITLE
Make /var/empty immutable (with chattr +i)

### PR DESCRIPTION
**Pull request by domenkozar**
See #18358 and #14910 what bugs these caused.

cc @edolstra

### CODE REVIEW

**Comment by edolstra** [comment]
No, don't make it a link to the Nix store! Paths in the Nix store can have group = `nixbld`, which will probably cause sshd to fail. Just do `chattr +i /var/empty`.

Also, `/var/empty` is not a GC root, which might cause problems.

**Comment by domenkozar** [comment]
@edolstra I fear that won't help much as `createHome` happens under root.

```
$ sudo su

$ mkdir test123

$ chattr -i test123

$ touch test123/a

$ ls test123
a
```

**Comment by domenkozar** [comment]
@edolstra updated

**Comment by domenkozar** [comment]
Just a side note: lots of these `mkdir -m` commands claim to be idempotent, but they're really not since permissions are not reset if directory exists.

**Comment by groxxda** [comment]
Could we use a tmpfs with `size=0,mode=000`?

**Comment by domenkozar** [comment]
@groxxda what advantage would that have over `chattr +i`?

**Comment by groxxda** [comment]
@domenkozar it's probably just me not being a fan of chattr :wink: 
only advantage I can think of is file-system support. But it's probably not relevant because nobody has `/var` on a `tmpfs` 

btw: Does your patch work when run twice?

**Comment by domenkozar** [comment]
@groxxda it did, but I pushed a fix for `rm: cannot remove '/var/empty': Operation not permitted`

**Comment by grahamc** [comment]
(Updated the title to reflect the solution)

**Comment by domenkozar** [comment]
@edolstra any objections?

File: nixos/modules/system/activation/activation-script.nix [hunk 1]
```diff
@@ -137,8 +140,13 @@ in
 
         mkdir -m 1777 -p /var/tmp
 
+        # Make sure it's really empty
+        chattr -i /var/empty
+        rm -rf /var/empty
```
Comment (edolstra) [inline nixos/modules/system/activation/activation-script.nix]: This introduces a race during activation where `/var/empty` doesn't exist. So for example, if it gets interrupted at that point, you won't be able to log in via ssh anymore.

File: nixos/modules/system/activation/activation-script.nix [hunk 2]
```diff
@@ -137,8 +140,13 @@ in
 
         mkdir -m 1777 -p /var/tmp
 
+        # Make sure it's really empty
+        chattr -i /var/empty
```
Comment (edolstra) [inline nixos/modules/system/activation/activation-script.nix]: This could be `${e2fsprogs}/bin/chattr` to prevent polluting the path with `e2fsprogs`.

**Comment by domenkozar** [comment]
@edolstra fixed

**Comment by domenkozar** [comment]
@fpletz I've linked two issues in description of the PR for motivation what issues we fix.

I think it's safe to disable this for containers.

**Comment by domenkozar** [comment]
@fpletz here needs to be added `chattr -R -f -i`: https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/virtualization/nixos-container/nixos-container.pl#L248

Any we probably should test if container deletion works in the tests.

**Comment by domenkozar** [comment]
Ah it already is, but doesn't fail: http://hydra.nixos.org/build/40856595/log/raw

**Pull request merged**